src/_pytest/capture.py

```diff
@@ -69,366 +165,357 @@ def pytest_load_initial_conftests(early_config: Config):
+                os.close(self.targetfd)
+            os.close(self.targetfd_invalid)
+        self.syscapture.done()
+        self.tmpfile.close()
+        self._state = "done"
 
-class EncodedFile(io.TextIOWrapper):
-    __slots__ = ()
+    def suspend(self):
+        self._assert_state("suspend", ("started", "suspended"))
+        if self._state == "suspended":
+            return
+        self.syscapture.suspend()
+        os.dup2(self.targetfd_save, self.targetfd)
+        self._state = "suspended"
 
-    @property
-    def name(self) -> str:
-        # Ensure that file.name is a string. Workaround for a Python bug
-        # fixed in >=3.7.4: https://bugs.python.org/issue36015
-        return repr(self.buffer)
+    def resume(self):
+        self._assert_state("resume", ("started", "suspended"))
+        if self._state == "started":
+            return
+        self.syscapture.resume()
+        os.dup2(self.tmpfile.fileno(), self.targetfd)
+        self._state = "started"
 
-    @property
-    def mode(self) -> str:
-        # TextIOWrapper doesn't expose a mode, but at least some of our
-        # tests check it.
-        return self.buffer.mode.replace("b", "")
+    def writeorg(self, data):
+        self._assert_state("writeorg", ("started", "suspended"))
+        """ write to original file descriptor. """
+        os.write(self.targetfd_save, data)
 
 
-CaptureResult = collections.namedtuple("CaptureResult", ["out", "err"])
+class FDCapture:
+    """Capture IO to/from a given os-level filedescriptor.
 
+    snap() produces text
+    """
 
-class MultiCapture:
-    out = err = in_ = None
-    _state = None
+    EMPTY_BUFFER = ""
+
+    def __init__(self, targetfd: int) -> None:
+        self._cap = FDCaptureBinary(targetfd)
+
+    def __repr__(self) -> str:
+        return self._cap.repr(self.__class__.__name__)
+
+    def start(self):
+        """ Start capturing on targetfd using memorized tmpfile. """
+        self._cap.start()
+
+    def snap(self):
+        # XXX use encoding of original stream
+        return self._cap.snap().decode("utf-8", errors="replace")
+
+    def done(self):
+        self._cap.done()
+
+    def suspend(self):
+        self._cap.suspend()
+
+    def resume(self):
+        self._cap.resume()
+
+    def writeorg(self, data):
+        """ write to original file descriptor. """
+        # XXX use encoding of original stream
+        self._cap.writeorg(data.encode("utf-8"))
+
+
+# MultiCapture
+
+CaptureResult = collections.namedtuple("CaptureResult", ["out", "err"])
+
+
+class MultiCapture:
+    _state = None
     _in_suspended = False
 
-    def __init__(self, out=True, err=True, in_=True, Capture=None):
-        if in_:
-            self.in_ = Capture(0)
-        if out:
-            self.out = Capture(1)
-        if err:
-            self.err = Capture(2)
+    def __init__(self, in_, out, err) -> None:
+        self.in_ = in_
+        self.out = out
+        self.err = err
```

Comment:
its great this logic is gone from the ctor, an we move the outside factory functions to a named constructor
to encapsulate the in/out/err  logic

also it looks to me if attrs couldbe used to declare the ctor/non-ctor atributes now

Reply:
attrs might be nice for some classes here. I'll try to remember this for later.